pip/req/req_set.py

```diff
@@ -432,8 +458,18 @@ def prepare_files(self, finder):
                         )
 
                 if not self.ignore_dependencies:
-                    for subreq in dist.requires(
+
+                    for missing in RequirementSet._missing_extras(
+                            dist.extras,
                             req_to_install.extras):
+                        logger.warning(
+                            '%s does not provide the extra \'%s\'',
+                            str(dist), missing)
+
+                    for subreq in dist.requires(
+                            RequirementSet._available_extras(
+                                dist.extras,
+                                req_to_install.extras)):
```

Comment:
@xavfernandez - Though, there is an issue with sets not guaranteeing the order of their items. Would a decent compromise be to just inline the comprehensions?